psi4/src/psi4/libmints/molecule.cc

```diff
@@ -2835,7 +2837,19 @@ Vector3 Molecule::fxyz(int atom) const { return input_units_to_au_ * full_atoms_
 
 double Molecule::xyz(int atom, int _xyz) const { return input_units_to_au_ * atoms_[atom]->compute()[_xyz]; }
 
-const double &Molecule::Z(int atom) const { return atoms_[atom]->Z(); }
+const double &Molecule::Z(int atom, bool zero_ghost) const {
+    Element_to_Z z_list = Element_to_Z();
+    if (zero_ghost) {
+        return atoms_[atom]->Z();
+    }
+    else if (atoms_[atom]->Z() == 0.0) {
+        static const double val = z_list[atoms_[atom]->symbol()];
+        return val;
+    }
+    else {
+        return atoms_[atom]->Z();
+    }
+}
```

Comment:
Can't this all be
```
if (!zero_ghost && (atoms_[atom]->Z() == 0.0)) {
    Element_to_Z z_list = Element_to_Z();
    return z_list[atoms_[atom]->symbol()];
}
else {
    return atoms_[atom]->Z();
}
```

Reply:
Much nicer! I agree on all accounts except `return z_list[atoms_[atom]->symbol()];`. This throws a warning at compile-time: 
```
/Users/mash/Gits/psi4/psi4/src/psi4/libmints/molecule.cc:2843:16: warning: returning reference to local temporary object
      [-Wreturn-stack-address]
        return z_list[atoms_[atom]->symbol()];
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
And I end up getting garbage for this return (Z = 6.95321009529437e-310 for Gh(He), type A and B).

Reply:
ok, I don't really know my C++ catechism. :-)

Reply:
I'm not 100% sure on the reason why, but declaring the type definitely fixes the problem. Someone with far more C++ knowledge than I have may be able to explain it! ( @amjames  )

Reply:
So, thanks to @amjames I've added a bit more to the test case: the molecule now includes an H and an He atom. This is because he predicted a failure when returning Z values from one atom type to another due to the `static const double val`. Basically, whichever atom got through to the z_list part first would determine all other `val`'s returned for that molecule. The new test cases catch this when checking all Z values for a mixed-element molecule when `zero_ghost = False` and both atoms are ghosted. 

This can really only be fixed by changing the return type of `Molecule::Z` itself to just `const double Molecule::Z` instead of a reference. By doing this, I can now directly `return z_list[atoms_[atom]->symbol()]`, meaning I won't have to set it as `static const` (so it can be overwritten properly).